pints/_mcmc/_smc.py

```diff
@@ -0,0 +1,266 @@
+#
+# Uses the Python `pymc3` (and `theano`) module to run SMC routine sampling.
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+import pints
+import numpy as np
+import theano
+import theano.tensor as tt
+from theano.compile.ops import as_op
+import pymc3 as pm
+# from pymc3.step_methods import smc
+
+
+# Set some flags for theano
+THEANO_FLAGS = 'floatX=float32,optimizer=fast_compile'
+theano.config.floatX = 'float64'
+theano.config.compute_test_value = 'raise'
+theano.config.exception_verbosity = 'high'
+
+
+#TODO: issue due to @as_op cannot be done inside class
+global evaluate, times
+
+
+class SMC(pints.MCMC):
+    """
+    *Extends:* :class:`MCMC`
+
+    Creates a chain of samples from a target distribution, using the SMC [1]
+    routine implemented in the `pymc3` module [2] with `theano` backend [3].
+
+    SMC stands for Sequential Monte Carlo.
+
+    [1] Ching, Chen, "Transitional Markov Chain Monte Carlo method for Bayesian
+    model updating, model class selection, and model averaging", Journal of
+    engineering mechanics, vol. 133, no. 7, p. 816-832, 2007.
+
+    [2] Patil, Huard, Fonnesbeck, "PyMC: Bayesian stochastic modelling in
+    Python", Journl of statistical software, vol. 35, no. 4, p. 1, 2010.
+
+    [3] Bergstra, Breuleux, Bastien, Lamblin, Pascanu, Desjardins, Turian,
+    Warde-Farley, Bengio, "Theano: A CPU and GPU Math compiler in Python", in
+    Proc. 9th Python in Science Conf., p. 1-7, 2010.
+
+    """
+    def __init__(self, likelihood, x0, sigma0=None):
+        super(SMC, self).__init__(likelihood, x0, sigma0)
+        self._problem = likelihood.problem()
+        self._evaluate = self._problem.evaluate
+
+        # pass the following to the _tt_evaluate functions
+        # TODO: @as_op cannot be done inside class
+        global evaluate, times
+        evaluate = self._evaluate
+        times = self._problem.times()
+
+        # Tensorflow wrapper
+        if isinstance(self._problem, pints.SingleSeriesProblem):
+            # For single series problem
+            self._tt_evaluate = _tt_evaluate_single_series
+            self._values = np.array([self._problem.values()]).T
+        else:
+            # For multiple series problem
+            raise NotImplementedError
+
+        # Check problem assumption:
+        # Guassian log_likelihood and should have prior
+        if not isinstance(self._log_likelihood, pints.BayesianLogLikelihood):
+            raise NotImplementedError(
+                'Currently only support pints.BayesianLogLikelihood')
+
+        # Marginal Likelihood
+        self._marginal_likelihood = None
+
+    def _get_prior_as_pm(self, prior):
+        """
+        Return PyMC3 type prior.
+        """
+        if isinstance(prior, pints.UniformPrior):
```

Comment:
Is this the only way to solve this? Would rather not test what type of prior is given! (or even _if_ it's a prior that's given, as long as it works it works (duck typing!))